src/backend/distributed/commands/alter_table.c

```diff
@@ -441,7 +472,16 @@ AlterDistributedTable(TableConversionParameters *params)
 		ereport(DEBUG1, (errmsg("setting multi shard modify mode to sequential")));
 		SetLocalMultiShardModifyModeToSequential();
 	}
-	return ConvertTable(con);
+
+	TableConversionReturn *tableConversionReturn = ConvertTable(con);
```

Comment:
I think the main problem is `ConvertTable`, so we should only take that into a mem. context, not the whole AlterDistributedTable/AlterTableSetAccessMethod/UndistributeTable etc.

Instead of copy & pasting the same pattern to all over, can we create a single function:
```python

def ConvertTable:
         create mem context
         return = ConvertTableInternal (rename the current ConvertTable to ConvertTableInternal)
         switch to original context
         copyConversion
         delete mem context
         return copyConversion
```

that could make this PR easier to review & merge

Reply:
We call AlterDistributedTable, UndistributeTable for each partition and colocated table during alteration of a table. I also wanted to make sure we are freeing memory allocated there. We allocate memory there for foreign key dependecy check.
But, let me check if we can leave it as it is by moving memory context creation-deletion to ConvertTable.

Reply:
I think that having a context for AlterDistributedTable, UndistributeTable is a safer approach even if we should repeat the pattern for both. If we also should handle partition related allocations, we can hanldle them in a nested partition context as always. 
e.g.
```c
MemoryContext citusPartitionContext =
			AllocSetContextCreate(CurrentMemoryContext,
								  "citus_per_partition_context",
								  ALLOCSET_DEFAULT_SIZES);
```

Reply:
I refactored it such that we free context only at ConvertTable. Memory allocations at AlterDistributedTable, UndistributeTable functions itself seems negligible.

Reply:
> I refactored it such that we free context only at ConvertTable

The current code is slightly complex. We create/switch many contexts. Why not [this](https://github.com/citusdata/citus/pull/6726#discussion_r1116857088)?

Reply:
(as Jelte's comment below shows, this is error prone style)